core/src/main/java/tc/oc/pgm/db/SQLDatastore.java

```diff
@@ -0,0 +1,307 @@
+package tc.oc.pgm.db;
+
+import static tc.oc.pgm.db.SettingsImpl.bitSettings;
+
+import java.io.File;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.time.Duration;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import tc.oc.pgm.api.Datastore;
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.map.MapActivity;
+import tc.oc.pgm.api.player.Username;
+import tc.oc.pgm.api.setting.SettingKey;
+import tc.oc.pgm.api.setting.SettingValue;
+import tc.oc.pgm.api.setting.Settings;
+import tc.oc.pgm.util.ClassLogger;
+
+public class SQLDatastore implements Datastore {
+
+  private final Logger logger;
+  private final Connection connection;
+  private final ExecutorService executorService;
+
+  public SQLDatastore(@Nullable String uri, ExecutorService executorService) throws SQLException {
+    this.logger = ClassLogger.get(PGM.get().getLogger(), getClass());
+    this.executorService = executorService;
+    this.connection = getConnection(uri);
+
+    initUsername();
+    initSettings();
+    initMapActivity();
+  }
+
+  private static Connection getConnection(@Nullable String uri) throws SQLException {
+    if (uri == null || uri.isEmpty()) {
+      uri = "sqlite:" + new File(PGM.get().getDataFolder(), "pgm.db").getAbsolutePath();
+    }
+
+    if (!uri.startsWith("jdbc:")) {
+      uri = "jdbc:" + uri;
+    }
+
+    try {
+      if (uri.startsWith("jdbc:sqlite")) {
+        Class.forName("org.sqlite.JDBC");
+      } else if (uri.startsWith("jdbc:mysql")) {
+        Class.forName("com.mysql.jdbc.Driver");
+      }
+    } catch (ClassNotFoundException e) {
+      throw new SQLException(
+          "Unable to load database driver class (your server likely does not support your database)");
+    }
+
+    final Connection connection = DriverManager.getConnection(uri);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *connection*. The resource is referenced in statements at the following line: 64. The resource is returned at line: 66. There are other execution paths that neither close the resource nor return it, e.g., when exception is thrown by Connection.setAutoCommit. To prevent this resource leak, close **connection** along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).